contracts/usul/BaseTokenVoting.sol

```diff
@@ -0,0 +1,172 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts-upgradeable/utils/cryptography/draft-EIP712Upgradeable.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+import "./BaseStrategy.sol";
+
+/// @title An abstract contract used as a base for ERC-20 token voting strategies
+abstract contract BaseTokenVoting is BaseStrategy {
+    enum VoteType {
+        Against,
+        For,
+        Abstain
+    }
+
+    struct ProposalVoting {
+        uint256 yesVotes; // The total number of YES votes for this proposal
+        uint256 noVotes; // The total number of NO votes for this proposal
+        uint256 abstainVotes; // The total number of ABSTAIN votes for this proposal
+        uint256 deadline; // The timestamp voting ends for this proposal
+        uint256 startBlock; // The block the proposal voting starts
+        mapping(address => bool) hasVoted;
+    }
+
+    uint256 public votingPeriod; // the length of time voting is valid for a proposal
+    uint256 public timelockPeriod;
+    string public name;
+
+    mapping(uint256 => ProposalVoting) internal proposals;
+
+    event TimelockPeriodUpdated(uint256 newTimeLockPeriod);
+    event VotingPeriodUpdated(uint256 newVotingPeriod);
+    event ProposalReceived(uint256 proposalId, uint256 timestamp);
+    event VoteFinalized(uint256 proposalId, uint256 timestamp);
+    event Voted(
+        address voter,
+        uint256 proposalId,
+        uint8 support,
+        uint256 weight
+    );
+
+    /// @notice Updates the voting time period
+    /// @param _newVotingPeriod The voting time period in seconds
+    function updateVotingPeriod(uint256 _newVotingPeriod) external onlyOwner {
+        _updateVotingPeriod(_newVotingPeriod);
```

Comment:
what is the purpose of having these external functions call internal functions here? couldn't we just have the login in the function?

Reply:
So it just DRYs up the code a bit. These internal functions can be used by the voting contracts that inherit from `BaseTokenVoting`. For example, `LinearTokenVoting` calls both `updateVotingPeriod` and `updateTimelockPeriod` from within the `setUp` function